File: ios/Shared/Sources/Metrics.swift
import Foundation

#if os(iOS)
    import UIKit
#endif

/// Where measurements go, if anywhere.
///
/// Beside the server address in Settings, because it is the same kind of decision made
/// by the same kind of person: somebody who runs their own things and wants to see how
/// they are doing. It defaults to nowhere, and nowhere is a complete answer.
enum MetricsSettings {
    private static let endpointKey = "metrics.endpoint"
    private static let headersKey = "metrics.headers"
    private static let instanceKey = "metrics.instance"

    /// The collector's metrics URL — the full one, ending in `/v1/metrics`.
    ///
    /// Stored as typed rather than as a host with a path appended here: collectors sit
    /// behind gateways, on odd ports, under prefixes, and guessing at the path is how
    /// somebody ends up with a 404 they cannot see and a setting that looks fine.
    static var endpoint: URL? {
        get {
            guard let raw = UserDefaults.standard.string(forKey: endpointKey),
                  !raw.isEmpty,
                  let url = URL(string: raw),
                  url.scheme == "https" || url.scheme == "http"
            else { return nil }
            return url
        }
        set {
            UserDefaults.standard.set(newValue?.absoluteString ?? "", forKey: endpointKey)
            NotificationCenter.default.post(name: .diagnosticsChanged, object: nil)
        }
    }

    /// What somebody typed, kept as typed so the settings field can show it back.
    ///
    /// One `Name: value` per line. Parsed on the way out rather than on the way in, so a
    /// half-typed header is a header that does nothing rather than a field that refuses
    /// to hold what is being typed into it.
    static var rawHeaders: String {
        get { UserDefaults.standard.string(forKey: headersKey) ?? "" }
        set {
            UserDefaults.standard.set(newValue, forKey: headersKey)
            NotificationCenter.default.post(name: .diagnosticsChanged, object: nil)
        }
    }

    static var headers: [String: String] {
        var found: [String: String] = [:]
        for line in rawHeaders.split(whereSeparator: \.isNewline) {
            let parts = line.split(separator: ":", maxSplits: 1)
            guard parts.count == 2 else { continue }
            let name = parts[0].trimmingCharacters(in: .whitespaces)
            let value = parts[1].trimmingCharacters(in: .whitespaces)
            guard !name.isEmpty, !value.isEmpty else { continue }
            found[name] = value
        }
        return found
    }

    /// Whether anything is measured at all.
    ///
    /// **Two conditions, and the first is not negotiable.** A device answering for
    /// itself reports nothing, ever — `Capabilities.syncing` is the same question asked
    /// of the screens, and it is false exactly when there is no far end. Somebody who
    /// chose to keep their shopping on one phone did not choose to have that phone
    /// describe itself to a third party, however anonymously, and an app whose whole
    /// premise is "your lists do not reach me" cannot have an exception in it. The
    /// second is that somebody has entered a collector.
    ///
    /// **Always false on the watch**, and for a reason that is about credentials rather
    /// than effort. A collector wants an endpoint and usually a header to authenticate
    /// with. The endpoint could reach the wrist the way the server's address does — in
    /// the application context — but the header cannot: a context is persisted and
    /// latest-wins, which is exactly why `WatchLink` refuses to put a session token in
    /// one. A watch reporting metrics would need a second credential channel, built for
    /// telemetry, on the device with the least to say. So the watch logs and does not
    /// measure, and what is worth knowing about the link (snapshots pushed, ticks
    /// replayed) is measured from the phone's end, where both are visible anyway.
    ///
    /// The calls are compiled on the watch all the same rather than fenced off with
    /// `#if`: `Store/Sources` and `Shared/Sources` are the same source in all three apps,
    /// and a platform check around every call site is how one of them comes to be
    /// written differently from the others.
    static var reporting: Bool {
        #if os(watchOS)
            return false
        #else
            return Capabilities.current.syncing && endpoint != nil
        #endif
    }

    /// A name for this installation, minted here and never sent anywhere else.
    ///
    /// OpenTelemetry wants a `service.instance.id` so that two devices reporting the
    /// same series can be told apart; without one, a phone and a Mac average into a
    /// meaningless middle. A random uuid rather than anything the system knows about
    /// this device: an identifier for advertising, a vendor id or a device name would
    /// each be a fact about somebody, and this is a fact about an install.
    static var instance: String {
        if let stored = UserDefaults.standard.string(forKey: instanceKey) { return stored }
        let minted = UUID().uuidString
        UserDefaults.standard.set(minted, forKey: instanceKey)
        return minted
    }
}

/// One attribute on a measurement.
///
/// The same ``Plain`` the log's `info` lines take, and that is the point: **a metric
/// name, label or attribute may not carry personal data either**, and rather than
/// writing the rule down twice it is the same type enforcing it. There is no way to put
/// a list name into a series here for the same reason there is no way to put one into an
/// `info` line — the value cannot be a `String` and the name can only be a literal.
struct Tagged: Hashable, Sendable {
    let name: String
    let value: String

    init(_ name: StaticString, _ value: Plain) {
        self.name = String(describing: name)
        self.value = value.written
    }
}

/// What this app measures about itself, and where it sends it.
///
/// ## Why this is written out rather than taken as a dependency
///
/// The OpenTelemetry Swift SDK is a package graph — the API, the SDK, an exporter,
/// swift-protobuf and gRPC underneath it — and what is needed here is one POST of one
/// JSON body, on a timer, from three app targets that already share every line of their
/// plumbing. The SDK earns its size in a service with traces, spans, context
/// propagation and a dozen instrumented libraries; this app has fourteen counters and
/// two histograms and no traces at all. It would also be the first dependency in these
/// targets that is not GRDB, on a project whose stated promise is that nothing about
/// somebody's shopping leaves the machine they chose — and a telemetry SDK is precisely
/// the dependency where that promise wants the fewest moving parts nobody here has read.
///
/// The cost of writing it out is this file: the OTLP/HTTP JSON shape, delta temporality,
/// and a fixed set of histogram buckets. That shape is a stable, published protocol; if
/// it ever needs traces, the SDK is the right answer and this is a hundred and fifty
/// lines to delete.
///
/// ## What it collects
///
/// Nothing, until somebody has both a server and a collector — see
/// ``MetricsSettings/reporting``. Then: how long requests take and how they end, how
/// deep the queue is and what draining it achieved, whether the change stream is up,
/// when the app decides it is offline, what sync refused, and that the app launched.
final class Metrics: @unchecked Sendable {

    static let shared = Metrics()

    /// How often what has been collected is sent.
    ///
    /// A minute. Short enough that a phone put down mid-shop has reported what happened,
    /// long enough that the report is one request rather than a stream of them — this
    /// runs on a battery, and a radio woken every five seconds costs more than the thing
    /// being measured.
    static let interval: TimeInterval = 60

    private let lock = NSLock()
    private var counters: [Instrument: Int] = [:]
    private var gauges: [Instrument: Int] = [:]
    private var histograms: [Instrument: Histogram] = [:]
    private var since = Date()
    private var sending: Task<Void, Never>?

    /// A metric name with its attributes, which together name one series.
    struct Instrument: Hashable, Sendable {
        let name: String
        let attributes: [Tagged]
    }

    /// Where a duration falls, in milliseconds.
    ///
    /// Fixed bounds rather than exponential ones: the questions worth asking of this app
    /// are "did it feel instant", "did it feel slow" and "did it time out", and these
    /// bracket all three. Chosen once and left alone — changing bucket bounds rewrites
    /// history in every dashboard drawn from them.
    static let bounds: [Double] = [5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000]

    struct Histogram: Sendable {
        var count = 0
        var sum: Double = 0
        var buckets = [Int](repeating: 0, count: Metrics.bounds.count + 1)

        mutating func add(_ value: Double) {
            count += 1
            sum += value
            let bucket = Metrics.bounds.firstIndex { value <= $0 } ?? Metrics.bounds.count
            buckets[bucket] += 1
        }
    }

    // MARK: - Recording

    /// One more of something.
    func count(_ name: StaticString, _ attributes: Tagged..., by amount: Int = 1) {
        guard MetricsSettings.reporting else { return }
        let key = Instrument(name: String(describing: name), attributes: attributes)
        lock.withLock { counters[key, default: 0] += amount }
    }

    /// What something is *now*, replacing whatever it was.
    ///
    /// For depths and sizes, where the sum of the readings means nothing and the last
    /// one means everything. Queue depth is the example: adding up every reading of it
    /// would say a queue of one that never drained was a queue of six hundred.
    func gauge(_ name: StaticString, _ value: Int, _ attributes: Tagged...) {
        guard MetricsSettings.reporting else { return }
        let key = Instrument(name: String(describing: name), attributes: attributes)
        lock.withLock { gauges[key] = value }
    }

    /// How long something took.
    func observe(_ name: StaticString, milliseconds: Double, _ attributes: Tagged...) {
        guard MetricsSettings.reporting else { return }
        let key = Instrument(name: String(describing: name), attributes: attributes)
        lock.withLock { histograms[key, default: Histogram()].add(milliseconds) }
    }

    // MARK: - Sending

    /// Starts the timer, if there is anywhere to send to.
    ///
    /// Called at every composition root and safe to call again: choosing a collector in
    /// Settings calls it, and so does giving one up, which is what stops it.
    func start() {
        sending?.cancel()
        sending = nil
        guard MetricsSettings.reporting else {
            // Whatever was collected before somebody turned this off is thrown away
            // rather than held for a collector that may never be named again.
            lock.withLock {
                counters.removeAll()
                gauges.removeAll()
                histograms.removeAll()
            }
            return
        }

        sending = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(Metrics.interval))
                guard !Task.isCancelled else { return }
                await self?.flush()
            }
        }

        #if os(iOS)
            // A phone put in a pocket is a phone that may not come back before the
            // system suspends it, and a suspended app's timer does not fire. The last
            // minute of a shopping trip is the minute worth having.
            NotificationCenter.default.addObserver(
                forName: UIApplication.willResignActiveNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                Task { await self?.flush() }
            }
        #endif
    }

    /// Sends what has been collected, and forgets it.
    ///
    /// Delta rather than cumulative, which is the one real decision in this file.
    /// Cumulative means holding every series for the life of the process and restarting
    /// the counts on relaunch — a collector then has to guess where a reset happened,
    /// and an app that is launched and killed twenty times a day gives it twenty chances
    /// to guess wrong. Delta means each report stands alone, so a report that is dropped
    /// costs that minute and nothing after it.
    func flush() async {
        guard MetricsSettings.reporting, let endpoint = MetricsSettings.endpoint else { return }

        let taken: (counters: [Instrument: Int], gauges: [Instrument: Int], histograms: [Instrument: Histogram], from: Date) =
            lock.withLock {
                defer {
                    counters.removeAll()
                    histograms.removeAll()
                    since = Date()
                    // Gauges are *not* cleared: a depth of nought is a fact worth
                    // reporting every minute, and a series that stops arriving is
                    // indistinguishable from a device that has been switched off.
                }
                return (counters, gauges, histograms, since)
            }

        guard !taken.counters.isEmpty || !taken.gauges.isEmpty || !taken.histograms.isEmpty else {
            return
        }

        let body = OTLP.body(
            counters: taken.counters,
            gauges: taken.gauges,
            histograms: taken.histograms,
            from: taken.from,
            to: Date()
        )
        guard let encoded = try? JSONSerialization.data(withJSONObject: body) else { return }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        // Whatever the collector needs to believe this. Never logged: the whole reason
        // this field exists is that it holds a credential.
        for (name, value) in MetricsSettings.headers {
            request.setValue(value, forHTTPHeaderField: name)
        }
        request.httpBody = encoded
        request.timeoutInterval = 15

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            if !(200..<300).contains(status) {
                // Warned rather than retried. A collector refusing a report is a
                // configuration mistake somebody has to go and fix, and a client that
                // retries into it turns one wrong header into a request every minute for
                // as long as the app is installed.
                Log.warn(
                    .metrics, "the collector refused a report",
                    Detail("status", .count(status)),
                    Detail("series", .count(taken.counters.count + taken.gauges.count + taken.histograms.count))
                )
            }
        } catch {
            // Dropped. Measurements about a shop with no signal are the least important
            // thing in the queue behind them, and holding them would mean a second
            // outbox with none of the first one's care.
            Log.warn(
                .metrics, "could not report",
                Detail("why", .failure(Plain.Failure(error)))
            )
        }
    }
}

/// The names, in one place.
///
/// Written out as constants so that a name is chosen once rather than typed at each call
/// site, where a `queue.drained` and a `queue.drain` would become two series that each
/// tell half the story. Dotted and lowercase, which is what the OpenTelemetry naming
/// conventions ask for.
enum Measured {
    /// How long a request to the server took, by route class and outcome.
    static let requestDuration: StaticString = "shoppinglist.request.duration"
    /// How many requests, by route class and outcome. Derivable from the histogram's
    /// count, and kept separately because a failure that never returned has a duration
    /// nobody should read as a latency.
    static let requests: StaticString = "shoppinglist.request.count"
    /// How many changes are waiting to be sent, right now.
    static let queueDepth: StaticString = "shoppinglist.queue.depth"
    /// What a drain achieved, by result.
    static let queueDrained: StaticString = "shoppinglist.queue.drained"
    /// A change stream opening, dropping, or refusing to open.
    static let streamState: StaticString = "shoppinglist.stream.state"
    /// A nudge arriving on one, by kind.
    static let streamNudge: StaticString = "shoppinglist.stream.nudge"
    /// The app deciding the far end is or is not there.
    static let reachability: StaticString = "shoppinglist.reachability.changed"
    /// Something the server would not accept, or a tick against a row that had gone.
    static let syncRefused: StaticString = "shoppinglist.sync.refused"
    static let syncLost: StaticString = "shoppinglist.sync.lost"
    /// A snapshot handed to the watch, and what became of it.
    static let watchSnapshot: StaticString = "shoppinglist.watch.snapshot"
    /// A cache or database write that failed.
    static let storeFailed: StaticString = "shoppinglist.store.failed"
    /// A migration or a handover between this device's store and a server's.
    static let handover: StaticString = "shoppinglist.handover"
    /// The app starting, by mode.
    static let launch: StaticString = "shoppinglist.launch"
}

File: android/app/src/main/kotlin/com/cernauskas/shoppinglist/diagnostics/Metrics.kt
package com.cernauskas.shoppinglist.diagnostics

import com.cernauskas.shoppinglist.BuildConfig
import com.cernauskas.shoppinglist.data.Capabilities
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Numbers about this app, pushed to a collector somebody else runs.
 *
 * ## Only where there is a server
 *
 * A device answering for itself collects nothing and sends nothing, and the guard is
 * [Capabilities.syncing] rather than a preference. That is the honest reading of what
 * standalone *is*: there is no far end, so there is no latency, no queue, no stream and
 * nothing to be offline from — every measurement below is a measurement of a
 * relationship that does not exist. It is also the promise the app makes on the settings
 * screen, and a promise enforced at one point is a promise rather than a habit.
 *
 * Somebody who has a server has a machine of their own and an operator who is themselves
 * — see docs/self-hosting.md, S8 — so pushing to a collector they configured is telling
 * themselves about their own phone. It is off until they type an address.
 *
 * ## Why this is not an SDK
 *
 * OTLP/HTTP is a POST of a JSON document, and what is below is that document. The
 * OpenTelemetry Android SDK brings a metric pipeline, a span pipeline, a context
 * propagator and several megabytes of it, for an app that wants eight instruments and
 * has an HTTP client already. Delta temporality means nothing has to be reconciled
 * across a restart either, which is most of what an SDK is carrying.
 *
 * ## What may be in here
 *
 * The same rule as the log, for the same reason and by the same means: every metric name
 * is written in this file, and every attribute value is an enum constant. Nothing that
 * came from a person can reach a label — see [Route], which is where a path stops being
 * a path.
 */
object Metrics {

    // MARK: - What is being counted

    /** A metric and the attributes it was recorded under. */
    private data class Series(val metric: String, val attributes: List<Pair<String, String>>)

    /** Counts since the last push. Delta, so a push that fails costs one window rather
     * than corrupting a running total. */
    private val counts = LinkedHashMap<Series, Long>()

    /** The latest reading of something that is a level rather than a total. */
    private val levels = LinkedHashMap<Series, Long>()

    /** Durations, bucketed. */
    private val durations = LinkedHashMap<Series, Buckets>()

    private val lock = Any()

    /** Milliseconds, and chosen for what actually happens: a home server on the same
     * wifi answers in tens, one over the internet in hundreds, and a phone in a tunnel
     * gives up in tens of thousands. */
    private val BOUNDS = doubleArrayOf(5.0, 10.0, 25.0, 50.0, 100.0, 250.0, 500.0, 1_000.0, 2_500.0, 5_000.0, 10_000.0)

    private class Buckets {
        val counts = LongArray(BOUNDS.size + 1)
        var total = 0.0
        var seen = 0L

        fun record(millis: Long) {
            seen += 1
            total += millis
            var at = BOUNDS.indexOfFirst { millis <= it }
            if (at < 0) at = BOUNDS.size
            counts[at] += 1
        }
    }

    // MARK: - Recording

    /**
     * The one gate, and every recording goes through it.
     *
     * Here rather than at the call sites, for the reason `Outbox.queue` gives about its
     * own guard: a caller that has to remember is a caller that will not, and this one
     * is a promise rather than an optimisation.
     */
    private inline fun whenSyncing(work: () -> kotlin.Unit) {
        if (!Capabilities.current.syncing) return
        work()
    }

    private fun count(metric: String, attributes: List<Pair<String, String>> = emptyList(), by: Long = 1) =
        whenSyncing {
            synchronized(lock) {
                val series = Series(metric, attributes)
                counts[series] = (counts[series] ?: 0L) + by
            }
        }

    private fun level(metric: String, value: Long) = whenSyncing {
        synchronized(lock) { levels[Series(metric, emptyList())] = value }
    }

    /** The app was opened. */
    fun launched() = count("shopping.app.launches")

    /** One request, how long it took and what became of it. */
    fun request(route: Route, outcome: Outcome, millis: Long) = whenSyncing {
        val series = Series(
            "shopping.request.duration",
            listOf("route" to route.name.lowercase(), "outcome" to outcome.name.lowercase()),
        )
        synchronized(lock) { durations.getOrPut(series) { Buckets() }.record(millis) }
    }

    /** How much is queued, as a level rather than a total. */
    fun queueDepth(waiting: Int) = level("shopping.queue.depth", waiting.toLong())

    /**
     * What became of a drain.
     *
     * Counted by what happened to the operations rather than by the drain, because a
     * drain that sent forty and lost one is not the same event as one that lost forty.
     */
    fun drained(sent: Int, waiting: Int, lost: Int, refused: Boolean) = whenSyncing {
        if (sent > 0) count("shopping.queue.operations", listOf("outcome" to "sent"), sent.toLong())
        if (lost > 0) count("shopping.queue.operations", listOf("outcome" to "lost"), lost.toLong())
        if (refused) count("shopping.sync.refusals")
        count(
            "shopping.queue.drains",
            listOf("outcome" to if (waiting > 0 && sent == 0) "stuck" else "drained"),
        )
    }

    /** The change stream came up or went down — the health that matters most, because a
     * stream that dies silently is what stops a screen ever learning it is back. */
    fun stream(opened: Boolean) =
        count("shopping.stream.transitions", listOf("state" to if (opened) "opened" else "closed"))

    /** The far end went out of reach, or came back. */
    fun reachability(reachable: Boolean) =
        count(
            "shopping.offline.transitions",
            listOf("state" to if (reachable) "online" else "offline"),
        )

    // MARK: - Pushing

    private val client = OkHttpClient.Builder()
        // Short. A collector that is not answering must not hold a coroutine open long
        // enough for the next window to pile up behind it.
        .callTimeout(15, TimeUnit.SECONDS)
        .build()

    /**
     * Pushes every minute, for as long as the scope lives.
     *
     * Started from `Application`, which is the only place with a lifetime that matches:
     * a scope tied to a screen would stop counting whenever somebody backgrounded the
     * app, which is exactly when a queue is interesting.
     */
    fun pushEvery(scope: CoroutineScope, seconds: Long = 60) {
        scope.launch {
            while (isActive) {
                delay(seconds * 1_000)
                push()
            }
        }
    }

    /**
     * Sends what has accumulated, and forgets it whether or not it arrived.
     *
     * Delta temporality is what makes that acceptable: a failed push costs one window
     * and the next one is complete, where a retry queue would be a second outbox — an
     * unbounded one, holding numbers, on a phone.
     */
    suspend fun push() {
        if (!Capabilities.current.syncing) return
        val endpoint = DiagnosticsSettings.endpoint
        if (endpoint.isBlank()) return

        val document = drain() ?: return

        withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url(endpoint)
                .post(document.toString().toRequestBody(JSON))
                .apply {
                    DiagnosticsSettings.headerPairs().forEach { (name, value) ->
                        header(name, value)
                    }
                }
                .build()

            try {
                client.newCall(request).execute().use { response ->
                    if (response.isSuccessful) {
                        Diagnostics.info(Event.METRICS_PUSHED)
                    } else {
                        // The status and nothing else. A collector's error body says
                        // what it did not like about the document, and this app does
                        // not know what is in one of those.
                        Diagnostics.warn(
                            Event.METRICS_REFUSED,
                            Fact.of(Field.STATUS, response.code),
                        )
                    }
                }
            } catch (problem: IOException) {
                Diagnostics.warn(Event.METRICS_REFUSED, Fact.failure(problem))
            }
        }
    }

    /** Takes everything counted since the last time, leaving the counters empty. */
    private fun drain(): JsonObject? {
        val takenCounts: Map<Series, Long>
        val takenLevels: Map<Series, Long>
        val takenDurations: Map<Series, Buckets>

        synchronized(lock) {
            if (counts.isEmpty() && levels.isEmpty() && durations.isEmpty()) return null
            takenCounts = LinkedHashMap(counts).also { counts.clear() }
            // Levels are not deltas and are not cleared: a queue that is empty and stays
            // empty is a reading worth having every minute, not a gap in a graph.
            takenLevels = LinkedHashMap(levels)
            takenDurations = LinkedHashMap(durations).also { durations.clear() }
        }

        val now = System.currentTimeMillis() * 1_000_000
        val began = (windowBegan.takeIf { it != 0L } ?: now).also { windowBegan = now }

        val metrics = buildJsonArray {
            takenCounts.forEach { (series, value) ->
                add(sum(series, value, began, now))
            }
            takenLevels.forEach { (series, value) ->
                add(gauge(series, value, now))
            }
            takenDurations.forEach { (series, buckets) ->
                add(histogram(series, buckets, began, now))
            }
        }

        return document(metrics)
    }

    /** When the window being reported started, in nanoseconds. */
    @Volatile
    private var windowBegan: Long = 0

    // MARK: - The document
    //
    // OTLP/HTTP with a JSON body, which is a shape rather than a protocol: the fields
    // below are `opentelemetry.proto.metrics.v1` as JSON, and a collector that accepts
    // protobuf on the same port accepts this on `/v1/metrics` with the JSON content
    // type. Numbers that are 64-bit go as strings, which is the protobuf JSON mapping
    // and not a quirk of this.

    private fun document(metrics: JsonArray): JsonObject = buildJsonObject {
        putJsonArray("resourceMetrics") {
            addJsonObject {
                putJsonObject("resource") {
                    putJsonArray("attributes") {
                        add(attribute("service.name", "shopping-list-android"))
                        add(attribute("service.version", BuildConfig.VERSION_NAME))
                        // Deliberately nothing that identifies the install. An id here
                        // would separate one phone's series from another's, and it would
                        // also be the one label in this document that is about a person
                        // rather than about the software. Whoever runs the collector can
                        // tell their phones apart by the fact that they own them.
                    }
                }
                putJsonArray("scopeMetrics") {
                    addJsonObject {
                        putJsonObject("scope") { put("name", "com.cernauskas.shoppinglist") }
                        put("metrics", metrics)
                    }
                }
            }
        }
    }

    private fun attribute(key: String, value: String): JsonObject = buildJsonObject {
        put("key", key)
        putJsonObject("value") { put("stringValue", value) }
    }

    private fun attributesOf(series: Series): JsonArray = buildJsonArray {
        series.attributes.forEach { (key, value) -> add(attribute(key, value)) }
    }

    private fun sum(series: Series, value: Long, began: Long, now: Long): JsonObject =
        buildJsonObject {
            put("name", series.metric)
            put("unit", "1")
            putJsonObject("sum") {
                // 1 is DELTA. Cumulative would mean this app remembering totals across
                // restarts, which is a database for something a collector already does.
                put("aggregationTemporality", 1)
                put("isMonotonic", true)
                putJsonArray("dataPoints") {
                    addJsonObject {
                        put("attributes", attributesOf(series))
                        put("startTimeUnixNano", began.toString())
                        put("timeUnixNano", now.toString())
                        put("asInt", value.toString())
                    }
                }
            }
        }

    private fun gauge(series: Series, value: Long, now: Long): JsonObject = buildJsonObject {
        put("name", series.metric)
        put("unit", "1")
        putJsonObject("gauge") {
            putJsonArray("dataPoints") {
                addJsonObject {
                    put("attributes", attributesOf(series))
                    put("timeUnixNano", now.toString())
                    put("asInt", value.toString())
                }
            }
        }
    }

    private fun histogram(series: Series, buckets: Buckets, began: Long, now: Long): JsonObject =
        buildJsonObject {
            put("name", series.metric)
            put("unit", "ms")
            putJsonObject("histogram") {
                put("aggregationTemporality", 1)
                putJsonArray("dataPoints") {
                    addJsonObject {
                        put("attributes", attributesOf(series))
                        put("startTimeUnixNano", began.toString())
                        put("timeUnixNano", now.toString())
                        put("count", buckets.seen.toString())
                        put("sum", buckets.total)
                        putJsonArray("bucketCounts") {
                            buckets.counts.forEach { add(it.toString()) }
                        }
                        putJsonArray("explicitBounds") { BOUNDS.forEach { add(it) } }
                    }
                }
            }
        }

    private val JSON = "application/json".toMediaType()

    /**
     * How many series are waiting to be pushed.
     *
     * For the test that holds the standalone guard in place. "Nothing was collected" is
     * the whole promise a device answering for itself makes, and it is not checkable
     * from outside without this.
     */
    fun recorded(): Int = synchronized(lock) { counts.size + levels.size + durations.size }

    /** Everything counted so far, thrown away. For a test, and for somebody who has
     * just changed which collector this talks to. */
    fun forget() {
        synchronized(lock) {
            counts.clear()
            levels.clear()
            durations.clear()
        }
    }
}
